# Normalizer

## Description from sklearn-site:

Normalize samples individually to unit norm.

Each sample (i.e. each row of the data matrix) with at least one non zero component is rescaled independently of other samples so that its norm (l1 or l2) equals one.

This transformer is able to work both with dense numpy arrays and scipy.sparse matrix (use CSR format if you want to avoid the burden of a copy / conversion).

Scaling inputs to unit norms is a common operation for text classification or clustering for instance. For instance the dot product of two l2-normalized TF-IDF vectors is the cosine similarity of the vectors and is the base similarity metric for the Vector Space Model commonly used by the Information Retrieval community.

## True meaning of Mr.N will be in the end of page :]

### Imports

In [109]:
import pandas as pd

from sklearn.preprocessing import Normalizer
from sklearn.model_selection import train_test_split
from sklearn.datasets import load_breast_cancer

from sklearn.linear_model import LogisticRegression

### Fuction for easy display

In [110]:
def dfinfo(df):
    print('======\n= HEAD\n======')
    display(df.head(10))
    
    print('======\n= DESC\n======')
    display(df.describe())

### Load data

In [111]:
data = load_breast_cancer()

X = pd.DataFrame(data.data)
X.columns = data.feature_names.tolist()

Y = pd.DataFrame(data.target)
Y.columns = ['result']

### Info about X

In [112]:
dfinfo(X)

= HEAD


mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   
5        12.45         15.70           82.57      477.1          0.12780   
6        18.25         19.98          119.60     1040.0          0.09463   
7        13.71         20.83           90.20      577.9          0.11890   
8        13.00         21.82           87.50      519.8          0.12730   
9        12.46         24.04           83.97      475.9          0.11860   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760         0.30010              0.14710         0.2419   
1           0.07864         0.08690              0.07017         0.1812   
2           0.15990         0.19740              0.12790         0.2069   
3           0.28390         0.24140              0.10520         0.2597   
4           0.13280         0.19800              0.10430         0.1809   
5           0.17000         0.15780              0.08089         0.2087   
6           0.10900         0.11270              0.07400         0.1794   
7           0.16450         0.09366              0.05985         0.2196   
8           0.19320         0.18590              0.09353         0.2350   
9           0.23960         0.22730              0.08543         0.2030   

   mean fractal dimension           ...             worst radius  \
0                 0.07871           ...                    25.38   
1                 0.05667           ...                    24.99   
2                 0.05999           ...                    23.57   
3                 0.09744           ...                    14.91   
4                 0.05883           ...                    22.54   
5                 0.07613           ...                    15.47   
6                 0.05742           ...                    22.88   
7                 0.07451           ...                    17.06   
8                 0.07389           ...                    15.49   
9                 0.08243           ...                    15.09   

   worst texture  worst perimeter  worst area  worst smoothness  \
0          17.33           184.60      2019.0            0.1622   
1          23.41           158.80      1956.0            0.1238   
2          25.53           152.50      1709.0            0.1444   
3          26.50            98.87       567.7            0.2098   
4          16.67           152.20      1575.0            0.1374   
5          23.75           103.40       741.6            0.1791   
6          27.66           153.20      1606.0            0.1442   
7          28.14           110.60       897.0            0.1654   
8          30.73           106.20       739.3            0.1703   
9          40.68            97.65       711.4            0.1853   

   worst compactness  worst concavity  worst concave points  worst symmetry  \
0             0.6656           0.7119                0.2654          0.4601   
1             0.1866           0.2416                0.1860          0.2750   
2             0.4245           0.4504                0.2430          0.3613   
3             0.8663           0.6869                0.2575          0.6638   
4             0.2050           0.4000                0.1625          0.2364   
5             0.5249           0.5355                0.1741          0.3985   
6             0.2576           0.3784                0.1932          0.3063   
7             0.3682           0.2678                0.1556          0.3196   
8             0.5401           0.5390                0.2060          0.4378   
9             1.0580           1.1050                0.2210    

= DESC


mean radius  mean texture  mean perimeter    mean area  \
count   569.000000    569.000000      569.000000   569.000000   
mean     14.127292     19.289649       91.969033   654.889104   
std       3.524049      4.301036       24.298981   351.914129   
min       6.981000      9.710000       43.790000   143.500000   
25%      11.700000     16.170000       75.170000   420.300000   
50%      13.370000     18.840000       86.240000   551.100000   
75%      15.780000     21.800000      104.100000   782.700000   
max      28.110000     39.280000      188.500000  2501.000000   

       mean smoothness  mean compactness  mean concavity  mean concave points  \
count       569.000000        569.000000      569.000000           569.000000   
mean          0.096360          0.104341        0.088799             0.048919   
std           0.014064          0.052813        0.079720             0.038803   
min           0.052630          0.019380        0.000000             0.000000   
25%           0.086370          0.064920        0.029560             0.020310   
50%           0.095870          0.092630        0.061540             0.033500   
75%           0.105300          0.130400        0.130700             0.074000   
max           0.163400          0.345400        0.426800             0.201200   

       mean symmetry  mean fractal dimension           ...             \
count     569.000000              569.000000           ...              
mean        0.181162                0.062798           ...              
std         0.027414                0.007060           ...              
min         0.106000                0.049960           ...              
25%         0.161900                0.057700           ...              
50%         0.179200                0.061540           ...              
75%         0.195700                0.066120           ...              
max         0.304000                0.097440           ...              

       worst radius  worst texture  worst perimeter   worst area  \
count    569.000000     569.000000       569.000000   569.000000   
mean      16.269190      25.677223       107.261213   880.583128   
std        4.833242       6.146258        33.602542   569.356993   
min        7.930000      12.020000        50.410000   185.200000   
25%       13.010000      21.080000        84.110000   515.300000   
50%       14.970000      25.410000        97.660000   686.500000   
75%       18.790000      29.720000       125.400000  1084.000000   
max       36.040000      49.540000       251.200000  4254.000000   

       worst smoothness  worst compactness  worst concavity  \
count        569.000000         569.000000       569.000000   
mean           0.132369           0.254265         0.272188   
std            0.022832           0.157336         0.208624   
min            0.071170           0.027290         0.000000   
25%            0.116600           0.147200         0.114500   
50%            0.131300           0.211900         0.226700   
75%            0.146000           0.339100         0.382900   
max            0.222600           1.058000         1.252000   

       worst concave points  worst symmetry  worst fractal dimension  
count            569.000000      569.000000               569.000000  
mean               0.114606        0.290076                 0.083946  
std                0.065732        0.061867                 0.018061  
min                0.000000        0.156500                 0.055040  
25%                0.064930        0.250400                 0.071460  
50%                0.099930        0.282200                 0.080040  
75%                0.161400        0.317900                 0.092080  
max                0.291000        0.663800                 0.207500  

[8 rows x 30 columns]

### Separating

In [113]:
X_train, X_test, y_train, y_test = train_test_split(X, Y, random_state=0)

### LogisticRegression time

In [114]:
lor = LogisticRegression()
lor.fit(X_train, y_train.values.ravel())

print('Train score:\t{:.3f}'.format(lor.score(X_train, y_train)))
print('Test score:\t{:.3f}'.format(lor.score(X_test, y_test)))

Train score:	0.960
Test score:	0.958


### Normalizing

In [115]:
N = Normalizer().fit(X)
X_n = pd.DataFrame(N.transform(X))
X_n.columns = X.columns

dfinfo(X_n)

= HEAD


mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0     0.007925      0.004573        0.054099   0.440986         0.000052   
1     0.008666      0.007486        0.055988   0.558619         0.000036   
2     0.009367      0.010109        0.061842   0.572276         0.000052   
3     0.016325      0.029133        0.110899   0.551922         0.000204   
4     0.009883      0.006985        0.065808   0.631774         0.000049   
5     0.013945      0.017586        0.092486   0.534398         0.000143   
6     0.009483      0.010382        0.062147   0.540411         0.000049   
7     0.012712      0.019313        0.083631   0.535813         0.000110   
8     0.014200      0.023834        0.095577   0.567784         0.000139   
9     0.014365      0.027716        0.096808   0.548661         0.000137   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0          0.000122        0.000132             0.000065       0.000107   
1          0.000033        0.000037             0.000030       0.000076   
2          0.000076        0.000094             0.000061       0.000098   
3          0.000406        0.000345             0.000150       0.000371   
4          0.000065        0.000096             0.000051       0.000088   
5          0.000190        0.000177             0.000091       0.000234   
6          0.000057        0.000059             0.000038       0.000093   
7          0.000153        0.000087             0.000055       0.000204   
8          0.000211        0.000203             0.000102       0.000257   
9          0.000276        0.000262             0.000098       0.000234   

   mean fractal dimension           ...             worst radius  \
0                0.000035           ...                 0.011181   
1                0.000024           ...                 0.010528   
2                0.000029           ...                 0.011212   
3                0.000139           ...                 0.021314   
4                0.000029           ...                 0.010979   
5                0.000085           ...                 0.017328   
6                0.000030           ...                 0.011889   
7                0.000069           ...                 0.015818   
8                0.000081           ...                 0.016920   
9                0.000095           ...                 0.017397   

   worst texture  worst perimeter  worst area  worst smoothness  \
0       0.007635         0.081325    0.889462          0.000071   
1       0.009862         0.066899    0.824026          0.000052   
2       0.012145         0.072545    0.812984          0.000069   
3       0.037881         0.141333    0.811515          0.000300   
4       0.008120         0.074137    0.767189          0.000067   
5       0.026602         0.115818    0.830664          0.000201   
6       0.014373         0.079607    0.834520          0.000075   
7       0.026091         0.102545    0.831674          0.000153   
8       0.033567         0.116004    0.807547          0.000186   
9       0.046900         0.112580    0.820167          0.000214   

   worst compactness  worst concavity  worst concave points  worst symmetry  \
0           0.000293         0.000314              0.000117        0.000203   
1           0.000079         0.000102              0.000078        0.000116   
2           0.000202         0.000214              0.000116        0.000172   
3           0.001238         0.000982              0.000368        0.000949   
4           0.000100         0.000195              0.000079        0.000115   
5           0.000588         0.000600              0.000195        0.000446   
6           0.000134         0.000197              0.000100        0.000159   
7           0.000341         0.000248              0.000144        0.000296   
8           0.000590         0.000589              0.000225        0.000478   
9           0.001220         0.001274              0.000255    

= DESC


mean radius  mean texture  mean perimeter   mean area  mean smoothness  \
count   569.000000    569.000000      569.000000  569.000000       569.000000   
mean      0.014843      0.022010        0.096006    0.605084         0.000113   
std       0.004011      0.011077        0.024912    0.047241         0.000059   
min       0.005512      0.004568        0.036396    0.376233         0.000022   
25%       0.011690      0.014104        0.076533    0.582554         0.000072   
50%       0.015061      0.020352        0.097151    0.615219         0.000104   
75%       0.017354      0.026888        0.111015    0.639475         0.000141   
max       0.028470      0.086609        0.178585    0.697401         0.000477   

       mean compactness  mean concavity  mean concave points  mean symmetry  \
count        569.000000      569.000000           569.000000     569.000000   
mean           0.000110        0.000079             0.000042       0.000212   
std            0.000065        0.000073             0.000023       0.000111   
min            0.000024        0.000000             0.000000       0.000041   
25%            0.000068        0.000039             0.000027       0.000137   
50%            0.000094        0.000064             0.000038       0.000194   
75%            0.000133        0.000103             0.000054       0.000264   
max            0.000557        0.000826             0.000161       0.000787   

       mean fractal dimension           ...             worst radius  \
count              569.000000           ...               569.000000   
mean                 0.000076           ...                 0.016745   
std                  0.000043           ...                 0.004009   
min                  0.000011           ...                 0.007245   
25%                  0.000045           ...                 0.013843   
50%                  0.000069           ...                 0.016722   
75%                  0.000094           ...                 0.019225   
max                  0.000319           ...                 0.032340   

       worst texture  worst perimeter  worst area  worst smoothness  \
count     569.000000       569.000000  569.000000        569.000000   
mean        0.029162         0.109652    0.777287          0.000155   
std         0.014452         0.025127    0.037726          0.000083   
min         0.005154         0.050496    0.696047          0.000027   
25%         0.018349         0.092567    0.749906          0.000099   
50%         0.027027         0.109760    0.768914          0.000142   
75%         0.036616         0.125560    0.799809          0.000190   
max         0.108567         0.205583    0.921243          0.000646   

       worst compactness  worst concavity  worst concave points  \
count         569.000000       569.000000            569.000000   
mean            0.000262         0.000261              0.000108   
std             0.000175         0.000236              0.000053   
min             0.000042         0.000000              0.000000   
25%             0.000148         0.000139              0.000077   
50%             0.000222         0.000213              0.000102   
75%             0.000322         0.000320              0.000133   
max             0.001474         0.002974              0.000416   

       worst symmetry  worst fractal dimension  
count      569.000000               569.000000  
mean         0.000335                 0.000099  
std          0.000170                 0.000056  
min          0.000046                 0.000015  
25%          0.000223                 0.000061  
50%          0.000308                 0.000089  
75%          0.000422                 0.000120  
max          0.001196                 0.000445  

[8 rows x 30 columns]

### Last separating

In [116]:
X_train_n, X_test_n, y_train, y_test = train_test_split(X_n, Y, random_state=0)

### Result (worse)

In [117]:
lor = LogisticRegression()
lor.fit(X_train_n, y_train.values.ravel())

print('Train score:\t{:.3f}'.format(lor.score(X_train_n, y_train)))
print('Test score:\t{:.3f}'.format(lor.score(X_test_n, y_test)))

Train score:	0.772
Test score:	0.762


## True meaning of Normalizer

### For each row

# $$\sum{X`_i^2} = 1$$

### Example

In [118]:
ex = [[5, 0, 5, 0]]
ex_n = Normalizer().fit_transform(ex)
print(ex_n)

[[0.70710678 0.         0.70710678 0.        ]]


### We have array
# $$ [5; 0; 5; 0] $$

### Normalized array

# $$ [0.70710678; 0; 0.70710678; 0] $$

### Let's power it and sum

# $$0.70710678^2 + 0^2 + 0.70710678^2 + 0^2 = 1$$

# Have a good time!

![giphy](giphy.gif)